dev/benchmarks/macrobenchmarks/lib/src/web/recorder.dart

```diff
@@ -332,175 +406,168 @@ class _WidgetBuildRecorderHostState extends State<_WidgetBuildRecorderHost> {
-  final Duration averageDrawFrameDuration;
+  /// This data will be used to display cards in the Flutter Dashboard.
+  final Map<String, Timeseries> scoreData = <String, Timeseries>{};
 
-  /// The noise, as a fraction of [averageDrawFrameDuration], measure from the [frames].
-  final double drawFrameDurationNoise;
+  /// This data isn't displayed anywhere. It's stored for completeness purposes.
+  final Map<String, dynamic> extraData = <String, dynamic>{};
 
-  /// Frame metrics recorded during a single benchmark run.
-  final List<FrameMetrics> frames;
-
-  Map<String, dynamic> toJson() {
-    return <String, dynamic>{
-      'name': name,
-      'scoreKeys': <String>['averageDrawFrameDuration'],
-      'averageDrawFrameDuration': averageDrawFrameDuration.inMicroseconds,
-      'drawFrameDurationNoise': drawFrameDurationNoise,
-      'frames': frames
-          .map((FrameMetrics frameMetrics) => frameMetrics.toJson())
-          .toList(),
-    };
+  /// Invokes [callback] and records the duration of its execution under [key].
+  Duration record(String key, VoidCallback callback) {
+    final Duration duration = timeAction(callback);
+    addDataPoint(key, duration);
+    return duration;
   }
 
-  @override
-  String toString() {
-    return _formatToStringLines(<String>[
-      'benchmark: $name',
-      'averageDrawFrameDuration: ${averageDrawFrameDuration.inMicroseconds}μs',
-      'drawFrameDurationNoise: ${drawFrameDurationNoise * 100}%',
-      'frames:',
-      ...frames.expand((FrameMetrics frame) =>
-          '$frame\n'.split('\n').map((String line) => '- $line\n')),
-    ]);
+  void addDataPoint(String key, Duration duration) {
+    scoreData
+        .putIfAbsent(key, () => Timeseries())
+        .add(duration.inMicroseconds);
   }
-}
 
-/// Contains metrics for a single frame.
-class FrameMetrics {
-  FrameMetrics._({
-    @required this.drawFrameDuration,
-    @required this.sceneBuildDuration,
-    @required this.windowRenderDuration,
-  });
+  /// Decides whether the data collected so far is sufficient to stop, or
+  /// whether the benchmark should continue collecting more data.
+  ///
+  /// The signals used are sample size, noise, and duration.
+  ///
+  /// If any of the timeseries doesn't satisfy the noise requirements, this
+  /// method will return true (asking the benchmark to continue collecting
+  /// data).
+  bool shouldContinue() {
+    // If we haven't recorded anything yet, we don't wanna stop now.
+    if (scoreData.isEmpty) {
+      return true;
+    }
+
+    final List<bool> shouldContinueList = <bool>[];
+    for (final String key in scoreData.keys) {
+      final Timeseries timeseries = scoreData[key];
 
-  /// Total amount of time taken by [Window.onDrawFrame].
-  final Duration drawFrameDuration;
+      // Collect enough data points before considering to stop.
+      if (timeseries.count < _kMinSampleCount) {
+        shouldContinueList.add(true);
+        continue;
```

Comment:
`break`? Or even `return true`?

Reply:
That's what I did initially. There was a bug and it was hard to debug and fix. So I decided to go for consistency and ease of debugging.

Actually, now that I think about it. It is probably easier to read if I turn this into a `List.map`:

```dart
final Iterable<bool> shouldContinueList = scoreData.keys.map((String key) {
  final Timeseries timeseries = scoreData[key];

  // Collect enough data points before considering to stop.
  if (timeseries.count < _kMinSampleCount) {
    return true;
  }

  // Is it still too noisy?
  if (timeseries.noise > _kNoiseThreshold) {
    // If the timeseries has enough data, stop it, even if it's noisy under
    // the assumption that this benchmark is always noisy and there's nothing
    // we can do about it.
    if (timeseries.count > _kMaxSampleCount) {
      print(
        'WARNING: Noise of benchmark "$name.$key" did not converge below '
        '${_percent(_kNoiseThreshold)}. Stopping because it reached the '
        'maximum number of samples $_kMaxSampleCount. Noise level is '
        '${_percent(timeseries.noise)}.',
      );
      return false;
    } else {
      return true;
    }
  }

  print(
    'SUCCESS: Benchmark converged below ${_percent(_kNoiseThreshold)}. '
    'Noise level is ${_percent(timeseries.noise)}.',
  );
  return false;
});
```
What do you think? No more `continue` or `break`.